src/blk/aio/aio.cc

```diff
@@ -25,33 +25,43 @@ int aio_queue_t::submit_batch(aio_iter begin, aio_iter end,
   int r;
 
   aio_iter cur = begin;
-  struct aio_t *piocb[aios_size];
-  int left = 0;
-  while (cur != end) {
-    cur->priv = priv;
-    *(piocb+left) = &(*cur);
-    ++left;
-    ++cur;
-  }
-  ceph_assert(aios_size >= left);
+#if defined(HAVE_LIBAIO)
+  struct aio_t *piocb[max_iodepth];
+#endif
   int done = 0;
-  while (left > 0) {
+  int itemCount0 = 0; //used for LIBAIO only
+  int itemCount = 0;
+  while (cur != end || itemCount0 < itemCount) {
 #if defined(HAVE_LIBAIO)
-    r = io_submit(ctx, std::min(left, max_iodepth), (struct iocb**)(piocb + done));
+    while (cur != end && itemCount < max_iodepth) {
+      cur->priv = priv;
+      piocb[itemCount] = &(*cur);
+      ++itemCount;
+      ++cur;
+    }
+    int toSubmit = itemCount - itemCount0;
+    r = io_submit(ctx, toSubmit, (struct iocb**)(piocb + itemCount0));
+    if (r >= 0 && r < toSubmit) {
+      r = -EAGAIN;
+      itemCount0 += (r > 0 ? r : 0);
```

Comment:
Easy task for an optimizer: `itemCount0 += 0;`, `as r==-EAGAIN` :)

Reply:
I believed I've fixed that... Let me check.